### ISSUE TITLE
Architecture questions

### ISSUE THREAD

**Issue by saeidakbari** [buraksezer/olric#78]
Hello,
Thanks for your good project.

I am writing to you to enquire about the architecture of Olric which is, by the way, looks a promising project. As a developer, I would be very interested in participating in this program, and I would be grateful if you could give me some further details.

Firstly, As described in project documentation it seems Olric uses some method called consistent for finding the place of the data that it could be. I would appreciate it if you could tell me if this method distributes data fairly amount cluster meaning each node member in the cluster takes the same amount of data for maintaining performance and hot partition problem? If so, are there any guarantees?

Secondly, the way we should use the program is by first creating an object called DMap and place keys and values by different available methods. I would appreciate it if you could tell me the way that a key that belongs to a DMap would distribute?  Are keys distribute amount many cluster nodes or they just store in one partition as documentation called cluster owner? If so what’s happening in the case of having a big instance of DMap and many other small ones that could potentially destroy a specific member?

Finally, I would appreciate some information about the performance of the query function. Is it enough for use in a highly traffic MQTT broker which has a lot of request at the same time? Can it be the performance bottleneck?

**Comment by buraksezer** [comment]
Hi @saeidakbari-ir 

I'm sorry for my delayed response. 

> Thanks for your good project.

I'm happy to hear that. I hope Olric would be a good solution for your case.

> I am writing to you to enquire about the architecture of Olric which is, by the way, looks a promising project. As a developer, I would be very interested in participating in this program, and I would be grateful if you could give me some further details.

We are always open for new developers and any kind of contribution. Keep in touch.

> I would appreciate it if you could tell me if this method distributes data fairly amount cluster meaning each node member in the cluster takes the same amount of data for maintaining performance and hot partition problem? If so, are there any guarantees?

[buraksezer/consistent](https://github.com/buraksezer/consistent) implements a paper from Google with a few extras. You can find details in the documentation. consistent hash function distributes partitions among the cluster members almost evenly. This means that some of the members may have more partitions than others. It should look like that: PartitionCount: 271, Nodes: A, B, C. After distribution: A: 80, B: 85, C: 106. 

So some of the members has more partitions than other members but it shouldn't be a problem. Partition size should be relatively small(around 100MB). In addition that, your application should distribute its own load evenly. So I mean that if your application wants to access a key million times in a second, this may be a design problem. You should cache it locally.

> maintaining performance and hot partition problem

All requests to a partition are processed by the partition owner. So if one of your partitions is under extreme load, this may lead availability problems.

> Are keys distribute amount many cluster nodes or they just store in one partition as documentation called cluster owner? 

All operations on keys are done by the **partition owner.** A partition can only have one partition owner. All members in the cluster have some amount of partitions.

>  If so what’s happening in the case of having a big instance of DMap and many other small ones that could potentially destroy a specific member?

DMap size or count is unimportant. If a member goes down, some portion of the partitions is served from their backup, if there is any. Otherwise the keys on that partitions would be unavailable. 

> Finally, I would appreciate some information about the performance of the query function. Is it enough for use in a highly traffic MQTT broker which has a lot of request at the same time? 

Query function runs in O(n) time. So its performance depends on the key count.  

> Can it be the performance bottleneck?

Yes, it is. There is no secondary index for range queries.

Understanding PACELC principle is important to have insight about internals of Olric: https://github.com/buraksezer/olric#pacelc-theorem

Basically, Olric is a lightweight re-implementation of Hazelcast in Golang. So Hazelcast documentation can be very helpful. See this video: https://www.youtube.com/watch?v=7zU9b48GUGQ
